File: src/cli/velocity_painting.py
import sys
from PIL import Image
import numpy as np

def load_image(image_file):
    image = Image.open(image_file).convert('L')
    return np.array(image)

def load_gcode(gcode_file):
    with open(gcode_file, 'r') as file:
        return file.readlines()

def save_gcode(gcode_lines, output_file):
    with open(output_file, 'w') as file:
        file.writelines(gcode_lines)

def adjust_print_speed(gcode_lines, image_data, projection_mode, z_offset, min_speed, max_speed):
    for i, line in enumerate(gcode_lines):
        if line.startswith('G1'):
            intensity = get_intensity(line, image_data, projection_mode, z_offset)
            speed = min_speed + (max_speed - min_speed) * intensity
            gcode_lines[i] = adjust_speed_in_line(line, speed)
    return gcode_lines

def get_intensity(line, image_data, projection_mode, z_offset):
    # Implement the logic to get the intensity based on the projection mode and image data
    # This is a placeholder implementation
    return 0.5

def adjust_speed_in_line(line, speed):
    parts = line.split()
    for i, part in enumerate(parts):
        if part.startswith('F'):
            parts[i] = f'F{speed}'
            break
    return ' '.join(parts) + '\n'

def main():
    if len(sys.argv) != 6:
        print("Usage: python velocity_painting.py <projection_mode> <params> <image_file> <source_gcode> <output_gcode>")
        sys.exit(1)

    projection_mode = sys.argv[1]
    params = sys.argv[2]
    image_file = sys.argv[3]
    source_gcode = sys.argv[4]
    output_gcode = sys.argv[5]

    image_data = load_image(image_file)
    gcode_lines = load_gcode(source_gcode)

    min_speed = 1000  # Example value, adjust as needed
    max_speed = 3000  # Example value, adjust as needed
    z_offset = 0  # Example value, adjust as needed

    adjusted_gcode = adjust_print_speed(gcode_lines, image_data, projection_mode, z_offset, min_speed, max_speed)
    save_gcode(adjusted_gcode, output_gcode)

if __name__ == "__main__":
    main()

File: src/libslic3r/VelocityPainting.cpp
#include "VelocityPainting.hpp"
#include <cmath>
#include <vector>
#include <algorithm>

namespace Slic3r {

void VelocityPainting::adjust_print_speed(std::vector<GCodeLine>& gcode_lines, const std::vector<std::vector<int>>& image_data, int projection_mode, double z_offset, double min_speed, double max_speed) {
    for (auto& line : gcode_lines) {
        double intensity = get_intensity(line, image_data, projection_mode, z_offset);
        double speed = min_speed + (max_speed - min_speed) * intensity;
        line.set_speed(speed);
    }
}

double VelocityPainting::get_intensity(const GCodeLine& line, const std::vector<std::vector<int>>& image_data, int projection_mode, double z_offset) {
    // Implement the logic to get the intensity based on the projection mode and image data
    // This is a placeholder implementation
    return 0.5;
}

void VelocityPainting::process_gcode(std::vector<GCodeLine>& gcode_lines, const std::vector<std::vector<int>>& image_data, int projection_mode, double z_offset, double min_speed, double max_speed) {
    adjust_print_speed(gcode_lines, image_data, projection_mode, z_offset, min_speed, max_speed);
}

} // namespace Slic3r
